HLTrigger/Timer/plugins/ThroughputService.cc

```diff
@@ -1,55 +1,91 @@
 // C++ headers
 #include <algorithm>
 #include <chrono>
+#include <ctime>
 
 // boost headers
 #include <boost/format.hpp>
 
 // CMSSW headers
+#include "FWCore/ParameterSet/interface/EmptyGroupDescription.h"
+#include "FWCore/Utilities/interface/TimeOfDay.h"
 #include "DQMServices/Core/interface/DQMStore.h"
 #include "ThroughputService.h"
 
 // local headers
 #include "processor_model.h"
 
+namespace {
+
+  template <typename T, typename Rep, typename Period>
+  T& operator<<(T& out, std::chrono::duration<Rep, Period> const& duration) {
```

Comment:
@fwyzard About the compile error, I wonder why this function is templated on `out` instead of just using `std::ostream& out`?

Reply:
I tried
```diff
diff --git a/HLTrigger/Timer/plugins/ThroughputService.cc b/HLTrigger/Timer/plugins/ThroughputService.cc
index 624c95840c94..59ab4860b7b8 100644
--- a/HLTrigger/Timer/plugins/ThroughputService.cc
+++ b/HLTrigger/Timer/plugins/ThroughputService.cc
@@ -17,8 +17,8 @@
 
 namespace {
 
-  template <typename T, typename Rep, typename Period>
-  T& operator<<(T& out, std::chrono::duration<Rep, Period> const& duration) {
+  template <typename Rep, typename Period>
+  std::ostream& operator<<(std::ostream& out, std::chrono::duration<Rep, Period> const& duration) {
     uint64_t microseconds = std::chrono::duration_cast<std::chrono::microseconds>(duration).count();
     uint32_t s = microseconds / 1000000;
     uint32_t us = microseconds % 1000000;
```
but it doesn't compile:
```
In file included from /data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/ErrorObj.h:132,
                 from /data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/MessageSender.h:6,
                 from /data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/MessageLogger.h:135,
                 from /data/user/fwyzard/hlt/CMSSW_11_1_0_pre4/src/HLTrigger/Timer/plugins/ThroughputService.h:21,
                 from /data/user/fwyzard/hlt/CMSSW_11_1_0_pre4/src/HLTrigger/Timer/plugins/ThroughputService.cc:13:
/data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/ErrorObj.icc: In instantiation of 'edm::ErrorObj& edm::ErrorObj::opltlt(const T&) [with T = std::chrono::duration<long int, std::ratio<1, 1000000> >]':
/data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/ErrorObj.icc:43:22:   required from 'edm::ErrorObj& edm::operator<<(edm::ErrorObj&, const T&) [with T = std::chrono::duration<long int, std::ratio<1, 1000000> >]'
/data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/MessageSender.h:41:23:   required from 'edm::MessageSender& edm::MessageSender::operator<<(const T&) [with T = std::chrono::duration<long int, std::ratio<1, 1000000> >]'
/data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/MessageLogger.h:267:12:   required from 'edm::LogInfo& edm::LogInfo::operator<<(const T&) [with T = std::chrono::duration<long int, std::ratio<1, 1000000> >]'
/data/user/fwyzard/hlt/CMSSW_11_1_0_pre4/src/HLTrigger/Timer/plugins/ThroughputService.cc:145:99:   required from here
/data/cmssw/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_0_pre4/src/FWCore/MessageLogger/interface/ErrorObj.icc:26:10: error: no match for 'operator<<' (operand types are 'std::ostringstream' {aka 'std::__cxx11::basic_ostringstream<char>'} and 'const std::chrono::duration<long int, std::ratio<1, 1000000> >')
     myOs << t;
     ~~~~~^~~~
```

Reply:
It appears that if you put the overload into `namespace std` it compiles on both gcc and clang. I believe (without checking the exact details) this behavior is caused by the  "argument-dependent lookup" in `std::ostringstream::operator<<()` that leads to `(anonymous namespace)::operator<<(std::ostream&, std::chrono::duration<...> const&)` not being picked up.

Declaring such an overload in `std` sounds like a timebomb though, because C++20 adds the `operator<<(std::ostream&, std::chrono::duration<...> const&)` overload as well. If you want a specific formatting for the `std::chrono::duration`, one option would be to wrap the `std::chrono::duration` into a custom type, and define the `operator<<()` for that type.

Reply:
Thanks for the suggestion.

Then the easiest solution is probably to add a constructor from `std::chrono::duration` to `edm::TimeOfDay`, and use that explicitly.